Promote odoc errors in dune

rgrinberg opened the pull request:
@aantron it seems like the way dune is generating the index file isn't correct anymore.

rgrinberg committed:
Fix odoc errors in dune tests

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Force the latest version of odoc in travis

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

aantron commented on test/blackbox-tests/test-cases/github717-odoc-index/index.mld:
```diff
@@ -1 +1 @@
-Test index.
\ No newline at end of file
+{1 Test index}
```
If this is a title, it should probably have level 0.

rgrinberg replied:
done

aantron commented on test/blackbox-tests/test-cases/odoc-unique-mlds/run.t:
```diff
@@ -8,6 +8,8 @@ Duplicate mld's in the same scope
           odoc _doc/_html/root/Root_lib2/.dune-keep,_doc/_html/root/Root_lib2/index.html
           odoc _doc/_odoc/pkg/root/page-index.odoc
+  File "../../../_mlds/root/index.mld", line 3, characters 0-21:
+  '{2': heading level should be lower than top heading level '2'.
```
Where is the code that generates this `.mld` file?

rgrinberg replied:
You probably answered your own question, but it's here:

https://github.com/rgrinberg/jbuilder/blob/fix-odoc-errors/src/odoc.ml#L418-L442

aantron reviewed (commented):
This:

https://github.com/ocaml/dune/blob/bc77e6dcef1e12dcdded336cf6bc32681f2bbc26/src/odoc.ml#L424

probably needs to be changed to `{0` as well.

rgrinberg committed:
Fix generated index.mld

Use 0 heading for first line and 1 for subsequent lines

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Change title heading to 0

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
@aantron I've addressed your comments. I still get these warnings:

```
'{0': heading level should be lower than top heading level '0'.
```

So I think the generated index is still off.

aantron commented on src/odoc.ml:
```diff
@@ -421,7 +421,7 @@ module Gen (S : sig val sctx : SC.t end) = struct
       Lib_name.compare (Lib.name x) (Lib.name y))
     |> List.iter ~f:(fun (lib, modules) ->
-      Printf.bprintf b "{2 Library %s}\n" (Lib_name.to_string (Lib.name lib));
+      Printf.bprintf b "{0 Library %s}\n" (Lib_name.to_string (Lib.name lib));
```
Ah, sorry, I missed the `List.iter` above. The page needs a separate title, with `{0`, which I think Dune is not generating at all right now. Below that, all these section headings should be `{1`.

rgrinberg replied:
Done. See the last commit.

rgrinberg commented:
I'm wondering if this fix deserves a 1.8.2 release. It's a bit of an ugly warning to have for every generated index.mld

aantron commented:
Looks good.

I would say a quick release is worthwhile, if it is easy to do.
